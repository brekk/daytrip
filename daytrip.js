'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var torpor = require('torpor');
var fluture = require('fluture');
var ramda = require('ramda');
var precinct = require('precinct');
var fg = require('fast-glob');
var path = require('path');
require('xtrace');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var precinct__default = /*#__PURE__*/_interopDefaultLegacy(precinct);
var fg__default = /*#__PURE__*/_interopDefaultLegacy(fg);
var path__default = /*#__PURE__*/_interopDefaultLegacy(path);

const glob = input =>
  new fluture.Future((bad, good) => {
    fg__default['default'](input).catch(bad).then(good);
    /* istanbul ignore next */
    return function cancel() {}
  });

const preslash = filepath =>
  ramda.pipe(ramda.indexOf(path__default['default'].sep), ramda.slice(0, ramda.__, filepath))(filepath);

const patharray = ramda.memoizeWith(ramda.identity)(ramda.split(path__default['default'].sep));
ramda.pipe(
  patharray,
  ramda.slice(-2, Infinity),
  ramda.join(path__default['default'].sep)
);

const keysOf = ramda.curry((pr, raw) =>
  ramda.pipe(ramda.propOr({}, pr), ramda.keys)(raw)
);
const getAllDependencies = ramda.pipe(
  ramda.of,
  ramda.ap([
    keysOf('dependencies'),
    keysOf('devDependencies'),
    keysOf('peerDependencies')
  ]),
  ([s, d, p]) => s.concat(d).concat(p)
);

const isDependency = ramda.curry((pkg, dep) =>
  ramda.pipe(
    getAllDependencies,
    ramda.either(ramda.includes(dep), ramda.includes(preslash(dep)))
  )(pkg)
);

/*
const aliases = {
  components: '/Users/brekk/work/charly/superdesk-browser3/src/components',
  layout: '/Users/brekk/work/charly/superdesk-browser3/src/layout',
  api: '/Users/brekk/work/charly/superdesk-browser3/src/api',
  modules: '/Users/brekk/work/charly/superdesk-browser3/src/modules',
  pages: '/Users/brekk/work/charly/superdesk-browser3/src/pages',
  setup: '/Users/brekk/work/charly/superdesk-browser3/src/setup',
  services: '/Users/brekk/work/charly/superdesk-browser3/src/services',
  sharedComponents: '/Users/brekk/work/charly/superdesk-browser3/src/sharedComponents',
  utils: '/Users/brekk/work/charly/superdesk-browser3/src/utils',
  src: '/Users/brekk/work/charly/superdesk-browser3/src',
  assets: '/Users/brekk/work/charly/superdesk-browser3/assets'
}
*/
const getAliasedPathFrom = ramda.curry((aliases, x) => {
  const alias = aliases[preslash(x)] || false;
  if (alias) {
    return path__default['default'].join(alias, x)
  }
  return x
});
const getAliasedPath = ramda.curry((aliases, a, b) => {
  const aliasedName = getAliasedPathFrom(aliases);
  const relativeTo = aliasedName(a);
  const filepath = aliasedName(b);
  if (a !== relativeTo) {
    console.log('aliased a!', a, '->', relativeTo);
  }
  if (b !== filepath) {
    console.log('aliased b!', b, '->', filepath);
  }
  if (filepath.startsWith('.')) {
    return path__default['default'].join(
      path__default['default'].dirname(filepath),
      path__default['default'].parse(filepath).base
    )
  }
  return filepath
});

const daytrip = ramda.curry((aliases, pkg, dir) =>
  ramda.pipe(
    glob,
    ramda.map(
      ramda.map(file =>
        ramda.pipe(
          torpor.readFile(ramda.__, 'utf8'),
          ramda.map(
            ramda.pipe(
              precinct__default['default'],
              ramda.reduce(
                (agg, dep) => {
                  const { modules: m, dependencies: d } = agg;
                  const filepath = getAliasedPath(aliases, file, dep);
                  const isModule = isDependency(pkg, filepath);
                  const dependencies = !isModule ? d.concat(dep) : d;
                  const modules = isModule ? m.concat(dep) : m;
                  return ramda.mergeRight(agg, { modules, dependencies })
                },
                { modules: [], dependencies: [] }
              ),
              ramda.mergeRight({ file })
            )
          )
        )(file)
      )
    ),
    fluture.chain(fluture.parallel(100))
  )(`${dir}/**/*.*`)
);

const fork = ramda.curry((bad, good, future) =>
  fluture.fork(bad)(good)(future)
);
ramda.curry((raw, x) => [raw, x]);

exports.daytrip = daytrip;
exports.fork = fork;
