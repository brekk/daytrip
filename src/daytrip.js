import { readFile } from 'torpor'
import { chain, parallel } from 'fluture'
import { __ as $, curry, map, mergeRight, pipe, reduce } from 'ramda'
import precinct from 'precinct'
import { glob } from './files'
import { getAliasedPath, isDependency } from './dependencies'

export const daytrip = curry((aliases, pkg, dir) =>
  pipe(
    glob,
    map(
      map(file =>
        pipe(
          readFile($, 'utf8'),
          map(
            pipe(
              precinct,
              reduce(
                (agg, dep) => {
                  const { modules: m, dependencies: d } = agg
                  const filepath = getAliasedPath(aliases, file, dep)
                  const isModule = isDependency(pkg, filepath)
                  const dependencies = !isModule ? d.concat(dep) : d
                  const modules = isModule ? m.concat(dep) : m
                  return mergeRight(agg, { modules, dependencies })
                },
                { modules: [], dependencies: [] }
              ),
              mergeRight({ file })
            )
          )
        )(file)
      )
    ),
    chain(parallel(100))
  )(`${dir}/**/*.*`)
)

export default daytrip
