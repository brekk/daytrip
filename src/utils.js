import { fork as rawFork } from 'fluture'
import { curry } from 'ramda'

export const fork = curry((bad, good, future) =>
  rawFork(bad)(good)(future)
)
export const pretack = curry((raw, x) => [raw, x])
