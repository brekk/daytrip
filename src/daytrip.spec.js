import pkg from '../package.json'
import { daytrip } from './daytrip'
import { fork } from './utils'

test('daytrip', done => {
  fork(
    done,
    out => {
      expect(out).toMatchSnapshot()
      done()
    },
    daytrip({}, pkg, 'src')
  )
})
