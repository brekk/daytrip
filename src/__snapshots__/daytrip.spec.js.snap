// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`daytrip 1`] = `
Array [
  Object {
    "dependencies": Array [
      "./files",
      "./dependencies",
    ],
    "file": "src/daytrip.js",
    "modules": Array [
      "torpor",
      "fluture",
      "ramda",
      "precinct",
    ],
  },
  Object {
    "dependencies": Array [
      "../package.json",
      "./daytrip",
      "./utils",
    ],
    "file": "src/daytrip.spec.js",
    "modules": Array [],
  },
  Object {
    "dependencies": Array [
      "./string",
    ],
    "file": "src/dependencies.js",
    "modules": Array [
      "ramda",
    ],
  },
  Object {
    "dependencies": Array [],
    "file": "src/files.js",
    "modules": Array [
      "fast-glob",
      "fluture",
    ],
  },
  Object {
    "dependencies": Array [
      "./files",
      "./utils",
    ],
    "file": "src/files.spec.js",
    "modules": Array [],
  },
  Object {
    "dependencies": Array [
      "./daytrip",
      "./utils",
    ],
    "file": "src/index.js",
    "modules": Array [],
  },
  Object {
    "dependencies": Array [],
    "file": "src/string.js",
    "modules": Array [
      "ramda",
    ],
  },
  Object {
    "dependencies": Array [
      "./string",
    ],
    "file": "src/string.spec.js",
    "modules": Array [],
  },
  Object {
    "dependencies": Array [],
    "file": "src/utils.js",
    "modules": Array [
      "fluture",
      "ramda",
    ],
  },
  Object {
    "dependencies": Array [
      "./utils",
    ],
    "file": "src/utils.spec.js",
    "modules": Array [],
  },
  Object {
    "dependencies": Array [],
    "file": "src/__snapshots__/daytrip.spec.js.snap",
    "modules": Array [],
  },
  Object {
    "dependencies": Array [],
    "file": "src/__snapshots__/files.spec.js.snap",
    "modules": Array [],
  },
]
`;
